app/src/main/java/ee/ioc/phon/android/speechutils/ContinuousRawAudioRecorder.java

```diff
@@ -0,0 +1,215 @@
+            System.arraycopy(mRecording, potentialStartSample, buffer, 0, numOfSamplesToGoBack);
+        }
+        else {
+            if (!mRecordingBufferIsFullWithData) {
+                Log.i(LOG_FILTER + "Start sample in the recording is a negative one. The buffer did not pass one cycle yet. Copying from position: 0, " + getLength() + " bytes");
+                System.arraycopy(mRecording, 0, buffer, 0, getLength());
+            }
+            else {
+                // the potential start sample is out of the boundaries of the array to the negative side
+                potentialStartSample = Math.abs(potentialStartSample);
+                Log.i(LOG_FILTER + "Start sample in the recording is a negative one. The buffer passed at least one cycle. Copying from position: " + (mRecording.length - potentialStartSample) + ", " + potentialStartSample + " bytes and from position: 0, " + getLength() + " bytes");
+                System.arraycopy(mRecording, mRecording.length - potentialStartSample, buffer, 0, potentialStartSample);
+                System.arraycopy(mRecording, 0, buffer, potentialStartSample, getLength());
+            }
+        }
+
+        return buffer;
+    }
+
+    private byte[] pcmToWav(byte[] pcm) {
+        return AudioUtils.getRecordingAsWav(pcm, getSampleRate(), RESOLUTION_IN_BYTES, CHANNELS);
+    }
+
+    private byte[] createWavHeader(int pcmDataLength) {
+        return AudioUtils.getWavHeader(pcmDataLength, getSampleRate(), RESOLUTION_IN_BYTES, CHANNELS);
+    }
+
+    public void dumpBufferToWavFile(String wavFileFullPath) {
+        AudioUtils.saveWavToFile(wavFileFullPath, pcmToWav(consumeRecording()), false);
+    }
+
+    public void startRecording(final String wavFileFullPath) {
+        if (!mRecordingToFile.compareAndSet(false, true))
+            return;
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                // in case that someone stopped the recording before the thread could start
+                if (!mRecordingToFile.get())
+                    return;
+
+                int pcmDataLength = 0;
+                byte[] pcmData;
+                AtomicBoolean firstBuffer = new AtomicBoolean(true);
+
+                while(mRecordingToFile.get()) {
+
+                    // No need to endlessly poll the recording. It will work without the sleep well but
+                    // every 50ms is also good (keeps the CPU happier than without the sleep)
+                    try {
+                        Thread.sleep(50L);
+                    }
+                    catch (InterruptedException e) {
+                        mRecordingToFile.set(false);
+                        return;
+                    }
+
+                    pcmData = consumeRecording();
+                    if (pcmData == null || pcmData.length == 0)
+                        continue;
+
+                    pcmDataLength += pcmData.length;
```

Comment:
here, the first sample that is written to the wav file is written with the wav header (which is a constant size header) and the rest of the samples are appended to the file as raw pcm data. The re-writing of the wav header with the updated pcm buffer size will occur at the end of the reading loop (L.207)